dbms/src/Storages/StorageBuffer.cpp

```diff
@@ -148,56 +149,65 @@ BlockInputStreams StorageBuffer::read(
+                if (!dst_col.type->equals(*col.type))
                 {
                     LOG_WARNING(log, "Destination table " << backQuoteIfNeed(destination_database) << "." << backQuoteIfNeed(destination_table)
                         << " has different type of column " << backQuoteIfNeed(column_name) << " ("
-                        << destination->getColumn(column_name).type->getName() << " != " << getColumn(column_name).type->getName()
-                        << "). Data from destination table is converted.");
-                    struct_mismatch = true;
+                        << dst_col.type->getName() << " != " << col.type->getName() << "). Data from destination table are converted.");
+                    header_after_adding_defaults.getByName(column_name) = ColumnWithTypeAndName(dst_col.type, column_name);
                 }
             }
-            else
+
+            if (columns_intersection.empty())
             {
                 LOG_WARNING(log, "Destination table " << backQuoteIfNeed(destination_database) << "." << backQuoteIfNeed(destination_table)
-                    << " doesn't have column " << backQuoteIfNeed(column_name) << ". The default values are used.");
-                struct_mismatch = true;
+                    << " has no common columns with block in buffer. Block of data is skipped.");
             }
-        }
-
-        if (columns_intersection.empty())
-            LOG_WARNING(log, "Destination table " << backQuoteIfNeed(destination_database) << "." << backQuoteIfNeed(destination_table)
-                << " has no common columns with block in buffer. Block of data is skipped.");
-        else
-        {
-            streams_from_dst = destination->read(columns_intersection, query_info, context, processed_stage, max_block_size, num_streams);
-            for (auto & stream : streams_from_dst)
-                stream->addTableLock(destination_lock);
-        }
-
-        if (struct_mismatch && !streams_from_dst.empty())
-        {
-            /// Add streams to convert read blocks from the destination table.
-            auto header = getSampleBlock();
-            for (auto & stream_from_dst : streams_from_dst)
+            else
             {
-                stream_from_dst = std::make_shared<AddingDefaultBlockInputStream>(
-                            stream_from_dst, header, getColumns().defaults, context);
-                stream_from_dst = std::make_shared<ConvertingBlockInputStream>(
-                            context, stream_from_dst, header, ConvertingBlockInputStream::MatchColumnsMode::Name);
+                streams_from_dst = destination->read(columns_intersection, query_info, context, processed_stage, max_block_size, num_streams);
+                for (auto & stream : streams_from_dst)
+                {
+                    stream = std::make_shared<AddingDefaultBlockInputStream>(
+                                stream, header_after_adding_defaults, getColumns().defaults, context);
```

Comment:
The assertion failure in the DEBUG build was caused by passing `header` to the constructor of `AddingDefaultBlockInput` here. It wasn't correct because the class `AddingDefaultBlockInput` is designed only to add default columns, it cannot convert existing columns to another types. `AddingDefaultBlockInput` cannot reach `header` by itself, it can reach only some intermediate header (which I called `header_after_adding_defaults`).